### ISSUE TITLE
Phase 50.2: GradientMetaLearner — MAML, Reptile & Meta-Gradient Optimization

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#971]
### Summary
Implement the `GradientMetaLearner` module providing MAML (Model-Agnostic Meta-Learning), Reptile, and first-order meta-gradient optimization methods. This module learns model initializations that can rapidly adapt to new tasks in a few gradient steps.

### Motivation
Gradient-based meta-learning optimizes for initial parameters θ* such that a few gradient steps on a new task's support set yields strong generalization on the query set. MAML achieves this via second-order gradients through the inner-loop optimization; Reptile offers a first-order approximation with comparable performance and simpler implementation.

### Architecture

```
┌──────────────────────────────────────────────────────┐
│              GradientMetaLearner                     │
│                                                      │
│  ┌───────────────────────────────────────────────┐   │
│  │              Meta-Parameter Store              │   │
│  │  θ* = argmin_θ Σ_τ L(U_τ^k(θ), D_τ^query)   │   │
│  └──────────────────┬────────────────────────────┘   │
│                     │                                │
│  ┌──────────┐ ┌─────▼─────┐ ┌──────────────────┐    │
│  │  MAML    │ │  Reptile  │ │  FOMAML /        │    │
│  │  Engine  │ │  Engine   │ │  ANIL / BOIL     │    │
│  │          │ │           │ │                   │    │
│  │ 2nd-ord. │ │ 1st-ord.  │ │ first-order      │    │
│  │ Hessian- │ │ moving    │ │ variants         │    │
│  │ vector   │ │ average   │ │                   │    │
│  └────┬─────┘ └─────┬─────┘ └────────┬──────────┘   │
│       │             │                │               │
│  ┌────▼─────────────▼────────────────▼────────────┐  │
│  │            Inner-Loop Optimizer                 │  │
│  │  k steps of SGD on support set per task        │  │
│  │  ∇_θ L(f_θ(x_support), y_support)             │  │
│  └────────────────────────────────────────────────┘  │
│  ┌────────────────────────────────────────────────┐  │
│  │          Outer-Loop Meta-Gradient              │  │
│  │  ∇_θ Σ_τ L(f_{θ'_τ}(x_query), y_query)      │  │
│  └────────────────────────────────────────────────┘  │
└──────────────────────────────────────────────────────┘
```

### Core API

```python
class GradientMetaLearner:
    def __init__(self, model: nn.Module, config: GradientMetaConfig):
        """Initialize with base model and meta-learning configuration."""
    
    def meta_train_step(self, episodes: List[Episode]) -> MetaTrainResult:
        """Execute one outer-loop meta-training step over a batch of episodes."""
    
    def adapt(self, support_x: Tensor, support_y: Tensor,
             num_steps: int = 5) -> nn.Module:
        """Adapt the meta-learned model to a new task using support set."""
    
    def evaluate(self, episode: Episode, num_adapt_steps: int = 5) -> EvalResult:
        """Evaluate: adapt on support, measure accuracy on query."""
    
    def set_algorithm(self, algo: str = "maml") -> None:
        """Switch algorithm: 'maml', 'fomaml', 'reptile', 'anil', 'boil'."""

@dataclass
class GradientMetaConfig:
    inner_lr: float = 0.01           # inner-loop learning rate
    outer_lr: float = 0.001          # outer-loop (meta) learning rate
    inner_steps: int = 5             # gradient steps per task
    first_order: bool = False        # use first-order approximation
    algorithm: str = "maml"          # 'maml', 'fomaml', 'reptile', 'anil'
    meta_batch_size: int = 4         # tasks per meta-update
    reptile_epsilon: float = 0.1     # Reptile interpolation rate

@dataclass
class MetaTrainResult:
    outer_loss: float
    inner_losses: List[float]        # per-task inner losses
    query_accuracies: List[float]    # per-task query accuracies
    grad_norm: float
    meta_step: int
```

### Key Features
- **Full MAML** with Hessian-vector products for second-order gradients
- **First-Order MAML (FOMAML)** — drop second-order terms for efficiency
- **Reptile** — first-order meta-learning via parameter interpolation
- **ANIL** (Almost No Inner Loop) — freeze feature extractor, adapt head only
- **BOIL** (Body Only update in Inner Loop) — complementary to ANIL
- **Multi-step inner loop** with configurable learning rate schedules
- **Gradient clipping** and stability monitoring in outer loop
- **Mixed-precision meta-training** support for memory efficiency

### Acceptance Criteria
- [ ] 5-way 5-shot miniImageNet accuracy ≥ 63% (MAML baseline)
- [ ] 5-way 1-shot miniImageNet accuracy ≥ 48% (MAML baseline)
- [ ] Reptile matches FOMAML within 1% accuracy on benchmarks
- [ ] Inner-loop adaptation converges in ≤ 10 steps
- [ ] Memory-efficient: train with meta-batch ≥ 4 on single GPU (16GB)
- [ ] ANIL variant reduces inner-loop compute by ≥ 50% vs full MAML
- [ ] Unit tests with ≥ 95% coverage, integration tests with Phase 50.1

### Academic References
- Finn, Abbeel, & Levine (2017) — MAML foundational paper
- Nichol, Achiam, & Schulman (2018) — Reptile first-order meta-learning
- Raghu et al. (2020) — "Rapid Learning or Feature Reuse?" (ANIL analysis)
- Oh et al. (2021) — "BOIL: Towards Representation Change for Few-Shot Learning"
- Antoniou, Edwards, & Storkey (2019) — "How to Train Your MAML" (MAML++ improvements)

### Dependencies
- Phase 50.1 (TaskDistributionSampler) for episode generation
- Phase 48.1 (ElasticWeightConsolidator) for continual meta-learning integration
